### PR TITLE
End-to-end differentiable potentials and tutorial

### PR DESCRIPTION
**Pull request by rohskopf**
This PR introduces the `custom` calculator which takes a full neighbor list from LAMMPS using my recent PR for getting a full neighbor list with pair zero: https://github.com/lammps/lammps/pull/3471

With this full neigh list, we can implement our own custom descriptors in a solver, and then perform optimization by backpropagating all the way back to the positions extracted from LAMMPS. This is "end-to-end" differentiable potential optimization, and has the advantage of allowing optimization of hyperparameters associated with the descriptors, along with using any general architecture such as pairwise networks. 

We implement this with pairwise networks in the new `networks` solver, using Bessel function descriptors, with the process illustrated below:
image

This produces decent agreement with forces for the Ta example, but not as good as SNAP, because it's only a 2-body radial potential. Future PR will include combining these pairwise networks with many-body terms. 

This PR also includes new library features in `fitsnap3lib.tools`:

- LammpsTools for using potentials with LAMMPS and calculating errors on an external test set using energies/forces extracted from LAMMPS.
- DataframeTools for post-processing errors, and descriptor analysis such as t-SNE (thanks @ldwillia)
- TestTools for performing validation tests/checks, such as finite difference calculations on the new custom networks, to ensure validity of force gradients.  

A `tutorial.ipynb` is also included in the main directory, with link to Google Colab environment.

### CODE REVIEW

**Review by mitwood — APPROVED** [review]

File: fitsnap3lib/calculators/lammps_custom.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        neighlist = []
+        xneighs = []
+        for i in range(0,nl.size):
+            idx, nlist  = nl.get(i)
+            #print("\natom {} with ID {} has {} neighbors:".format(idx,tags[idx],nlist.size))
+            num_neighs_i = 0
+            if nlist.size > 0:
+                for n in np.nditer(nlist):
+                    num_neighs_i += 1
+                    neighlist.append([tags[idx], tags[n]]) #, ptr_pos[n][0], ptr_pos[n][1], ptr_pos[n][2]])
```
[the remark this answers is not in the document]
Comment (rohskopf) [reply fitsnap3lib/calculators/lammps_custom.py:124]: Didn't know that. I'll see if it's a problem here - that `neighlist` gets destroyed and recreated every configuration, so I don't think it'll be too much of a problem if our configurations are DFT sized. We save a copy in a shared array though.
